### DISCUSSION TITLE
How do I prevent a Commutator (or indeed any given subexpression species) from being evaluated by doit()? More generally, how do hints work?

### DISCUSSION THREAD

**Discussion by CharlieDerby** [sympy/sympy#25434] [Q&A] [answered]
I have a set of integrals involving commutators of various (constant) symbols. It's my preference to keep the commutators written as is when evaluating the integrals, how I can I keep them intact when I evaluate the expression? Here is a basic example of what I mean:

What I write
```
from sympy import *
from sympy.physics.quantum import Commutator

A,B = symbols('A B', commutative=False)
x = symbols('x')

expr = 1 + Integral(Commutator(A,B),(x,0,2))

expr.doit()
```
What I want:
```
>>> 1 + 2[A,B]
```
What I get:
```
>>> 1 + 2(AB-BA)
```

If I set `deep=False` I just get the unevaluated expression. I suspect that it is only evaluating the `Add` layer of the expression and stopping at the integral because if I try this on a single integral then it works. I could probably manually run through the expression and evaluate the individual integrals with `deep=False` but it would be simpler if I could just tell sympy to leave the commutators alone.

I see that the docs say that this should be possible via "hints": 

https://docs.sympy.org/latest/modules/core.html#sympy.core.basic.Basic.doit

_"sympy.core.basic.Basic.doit(\*\*hints)_

_Evaluate objects that are not evaluated by default like limits, integrals, sums and products. All objects of this kind will be evaluated recursively, **unless some species were excluded via ‘hints’** or unless the ‘deep’ hint was set to ‘False’."_

I've googled around and had a dig through the source code but it remains unclear to me how one would go about using hints to exclude a particular class of subexpression from being evaluated. 

Indeed it's unclear to me how the hints system works at all. From what I read in the docs it sounds like it's a potentially quite powerful and general system for tweaking exactly how sympy treats your expressions but it seems to be totally undocumented. Is this what it is? If so, can anyone tell me how it works?

**Comment by oscarbenjamin** [comment, marked as answer]
There was recent discussion about this in gh-25383.

Here is a simple way to get what you asked for:
```python
In [43]: from sympy import *
    ...: from sympy.physics.quantum import Commutator
    ...: 
    ...: A,B = symbols('A B', commutative=False)
    ...: x = symbols('x')
    ...: 
    ...: expr = 1 + Integral(Commutator(A,B),(x,0,2))

In [44]: expr
Out[44]: 
    2         
    ⌠         
1 + ⎮ [A,B] dx
    ⌡         
    0         

In [45]: expr.replace(lambda e: isinstance(e, Integral), lambda e: e.doit(deep=False))
Out[45]: 1 + 2⋅[A,B]
```

**Reply by CharlieDerby** [reply]
Thank you, this will do fine for my purposes.

Thanks for linking the earlier discussion. Do I understand correctly that currently there is no general system for excluding expression types from `doit` via hints and that this is being worked on? If that is the case then perhaps the documentation should be updated to reflect this to avoid confusions like mine. In the meantime, an example like the one you posted might be helpful as I'm sure a lot of people will want to do a similar move.

**Reply by oscarbenjamin** [reply]
It would be good for someone to work on it but I don't know that anyone is working on it. As I said in the PR a simple fix would be for someone to add a `doit` function that can do the equivalent of what I showed.
